### DISCUSSION TITLE
Consider updating Newtonsoft.Json

### DISCUSSION THREAD

**Discussion by Jdance-Media** [SmartlyDressedGames/Unturned-3.x-Community#4792] [Ideas]
<img width="428" alt="image" src="https://github.com/user-attachments/assets/743e46c1-6885-4550-9647-c66e44f45a46">

7.0.1 doesn't play nice with .NET Standard 2.1 and it is marked with a security vulnerability which results in Visual Studio being annoying about it. I feel like it would be beneficial to upgrade to v13 as it has more to work with and still targets .NET Framework.

**Comment by DanielWillett** [comment]
Kind of related, have you considered releasing official NuGet packages as part of your update workflow?

**Comment by SDGNelson** [comment]
I would if there's a way we could update the project from .NET Framework to .NET Standard without breaking compatibility? .NET Standard 2.1 is available as an option in Unity 2021 LTS.

@DanielWillett As in, of the Unturned Assembly-CSharp to NuGet?

**Reply by andriivitiv** [reply]
I'm pretty sure it's not a breaking change. All it does is prevent you from using APIs that are not available in .NET Standard (e.g. windows specific methods) which is a good thing and will prevent issues like https://github.com/SmartlyDressedGames/Unturned-3.x-Community/issues/3901. In the same way, you cannot currently use spans even though they are already included in mscorlib.dll.

This compatibility option is just an editor thing and shouldn't affect plugins. Assemblies in the `Managed` folder will remain the same (but you might want to double-check this) and a few new ones might be added.

If you have precompiled dlls targeting .NET Standard, Unity will likely show a "missing dependency" error. To resolve this, copy the necessary assemblies from `Editor/Data/MonoBleedingEdge/lib/mono/4.5/Facades/`. They include `TypeForwardedTo` attributes. OpenMod does this to support netstandard plugins.

**Reply by GazziFX** [reply]
NET Standard also remove dynamic method emit. Not sure if that breaks Harmony because it has NET Standard 2.0 tfm. And also Unturned list accessor need to be replaced with FieldInfo.GetValue

**Reply by DanielWillett** [reply]
You can still use dynamic code you just have to reference the emit packages, and harmony 2.3.3 has a netstandard2.1 target now.

**Reply by SDGNelson** [reply]
Thank you for the insights!

**Comment by DanielWillett** [comment]
Yeah and all the other stuff like Glazier, netpack, system and unity ex, etc, and steamworks if you can. OpenMod has a Redist one but it always takes a few weeks for it to get updated.


You should be able to upgrade without issues i think, you'd have to test to make sure but i think the only issue would be with types that used to be in a separate library but are now in netstandard like System.Span and System.ReadOnlySpan (which used to be in System.Memory) but you dont really expose anything with spans in it.

With that said the latest Newtonsoft.Json supports .NET framework 2.0+ so you wouldn't have to change anything.
![image](https://github.com/user-attachments/assets/abf3437e-7027-43a0-ade1-70ca6929df42)

**Reply by SDGNelson** [reply]
I don't think I've personally used NuGet before, or at least not knowingly. It sounds like it would be the more modern way to redistribute the game libraries for c# developers.

**Reply by Jdance-Media** [reply]
Yep. I've personally been using the packages provided by RocketModFix. NuGet is easier to work with rather than having a lib folder with the libraries in it. It would go a long way if SDG had NuGet packages that could be referenced. While RMF has automated the packages so they are up-to-date, it would still be best for an official package to be out there.

**Comment by Jdance-Media** [comment]
Yeah, my original request was more so to fix some of the warnings Visual Studio gives out while adding more functionality in regards to JSON. For instance, my code used some methods for JObject that do not exist in that ancient version. The package for Newtonsoft.JSON already supports .NET Framework, so I'm not sure if an upgrade to .NET Standard is needed in relation to that.

With that being said, plugins have full on support for netstandard. We simply just add the library netstandard.dll and any other libraries we are using in regard to that .NET version. I use .NET Standard 2.1 for modules and it works out perfectly fine so long as I include the library for it.

Considering we already use it for most new plugins, you might be able to support .NET Standard 2.1 pretty easily.

**Reply by SDGNelson** [reply]
For the next update at least I will update Json.NET assuming it doesn't have any negative side effects!
